### PR TITLE
Fix #967 by adding http.xxx attributes on grpc server_interceptor

### LINKED ISSUE #967: Request name is empty when export grpc service to Azure Moniter.
I'm trying to integrate our python application with Azure Application Insights.
The python application is used for providing grpc service.
So I followed the hello_word_server example from opencensus-ext-grpc but replace the stackdriver exporter with AzureExporter.

but the Request name is displayed as undefined in Application Insights.
Any clue what's the reason?

**Describe your environment.** 
Python 3.8
Ubuntu 20.04

**Steps to reproduce.**

```python
import time
from concurrent import futures

import grpc

import hello_world_pb2
import hello_world_pb2_grpc
from opencensus.ext.grpc import server_interceptor
# from opencensus.ext.stackdriver import trace_exporter as stackdriver_exporter
from opencensus.trace import samplers
from opencensus.ext.azure.trace_exporter import AzureExporter

_ONE_DAY_IN_SECONDS = 60 * 60 * 24

# Callback function to add os_type: linux to span properties
def callback_function(envelope):
    envelope.data.baseData.properties['os_type'] = 'linux'
    return True

class HelloWorld(hello_world_pb2_grpc.GreeterServicer):

    def SayHello(self, request, context):
        return hello_world_pb2.HelloReply(message='Hello, %s!' % request.name)


def serve():
    sampler = samplers.AlwaysOnSampler()
    # exporter = stackdriver_exporter.StackdriverExporter()
    exporter = AzureExporter()
    exporter.add_telemetry_processor(callback_function)

    tracer_interceptor = server_interceptor.OpenCensusServerInterceptor(
        sampler, exporter)
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10),
        interceptors=(tracer_interceptor,))
    hello_world_pb2_grpc.add_GreeterServicer_to_server(HelloWorld(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    try:
        while True:
            time.sleep(_ONE_DAY_IN_SECONDS)
    except KeyboardInterrupt:
        server.stop(0)


if __name__ == '__main__':
    serve()

```

**What is the expected behavior?**
Expect to see the function path as the name.
So in this case: helloworld.Greeter.SayHello

**What is the actual behavior?**
name is undefined/empty in the application insights GUI.

image

here is the envelope that I captured with callback:

```json
{
   "iKey":"619cfb06-af79-4759-9dac-9047b7b69cee",
   "tags":{
      "ai.cloud.role":"hello_world_server.py",
      "ai.cloud.roleInstance":"fan-NUC10i7FNK",
      "ai.device.id":"fan-NUC10i7FNK",
      "ai.device.locale":"en_US",
      "ai.device.osVersion":"#57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020",
      "ai.device.type":"Other",
      "ai.internal.sdkVersion":"py3.8.5:oc0.7.11:ext1.0.5",
      "ai.operation.id":"a6dd3500130b9f9ff5b824e042372886",
      "ai.operation.parentId":"0000000000000000"
   },
   "time":"2020-10-28T09:05:37.270290Z",
   "name":"Microsoft.ApplicationInsights.Request",
   "data":{
      "baseData":{
         "id":"9d62b242ef1c012b",
         "duration":"0.00:00:00.000",
         "responseCode":"0",
         "success":True,
         "properties":{
            "component":"grpc"
         },
         "ver":2,
         "name":"",
         "source":"None",
         "url":"None",
         "measurements":"None"
      },
      "baseType":"RequestData"
   },
   "ver":1,
   "sampleRate":"None",
   "seq":"None",
   "flags":"None"
}
```

### PR DESCRIPTION
**Pull request by fanshaohua-fan**
Fix for issue #967

### CODE REVIEW

File: contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py [hunk 1]
```diff
@@ -28,6 +28,12 @@
 ATTRIBUTE_COMPONENT = 'COMPONENT'
 ATTRIBUTE_ERROR_NAME = 'ERROR_NAME'
 ATTRIBUTE_ERROR_MESSAGE = 'ERROR_MESSAGE'
+ATTRIBUTE_HTTP_HOST = 'HTTP_HOST'
```
Comment (lzchen) [inline contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py:31]: You can use the `attributes.helper` provided in `trace` similarly to [flask[(https://github.com/census-instrumentation/opencensus-python/blob/master/contrib/opencensus-ext-flask/opencensus/ext/flask/flask_middleware.py#L36) instead.

File: contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py [hunk 2]
```diff
@@ -107,10 +113,36 @@ def _start_server_span(self, servicer_context):
...
+                ATTRIBUTE_HTTP_METHOD),
+            attribute_value='POST')
+        tracer.add_attribute_to_current_span(
+            attribute_key=attributes_helper.COMMON_ATTRIBUTES.get(
+                ATTRIBUTE_HTTP_ROUTE),
+            attribute_value=grpc_method)
+        tracer.add_attribute_to_current_span(
+            attribute_key=attributes_helper.COMMON_ATTRIBUTES.get(
+                ATTRIBUTE_HTTP_URL),
+            attribute_value=grpc_host + grpc_method)
```
Comment (lzchen) [inline contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py:146]: Add `STATUS_CODE` as well?

Comment (lzchen) [inline contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py:146]: Is there anything for `HTTP_PATH` that can be used?

**Comment by fanshaohua-fan** [comment]
@lzchen , code has been adjusted based on your reviews:
1. use attributes.helper instead
2. STATUS_CODE is added following this [mapping](https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto)
3. add HTTP_PATH with the same value of HTTP_ROUTE
4. add `grpc://` prefix for HTTP_URL

File: contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py [hunk 3]
```diff
@@ -107,28 +142,60 @@ def _start_server_span(self, servicer_context):
...
+        )
+        tracer.add_attribute_to_current_span(
+            GRPC_METHOD, grpc_method
+        )
+
+        tracer.add_attribute_to_current_span(
+            HTTP_HOST, grpc_host
+        )
+        tracer.add_attribute_to_current_span(
+            HTTP_METHOD, 'POST'
```
Comment (lzchen) [inline contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py:161]: Sorry not too familiar with grpc, what's the different between `grpc_method` and `method`? Why is `method` always POST?

File: contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py [hunk 4]
```diff
@@ -107,28 +142,60 @@ def _start_server_span(self, servicer_context):
...
+        )
+        tracer.add_attribute_to_current_span(
+            HTTP_ROUTE, grpc_method
```
Comment (lzchen) [inline contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py:164]: Why are the `ROUTE` and `PATH` set to `grpc_method`?

**Review by lzchen — APPROVED** [review]
Looks good. Is there anything grpc specific that we should be aware of? @aabmass

File: contrib/opencensus-ext-grpc/tests/test_server_interceptor.py [hunk 5]
```diff
@@ -110,7 +133,9 @@ def test_intercept_handler_exception(self):
                 None, None)
             mock_context = mock.Mock()
             mock_context.invocation_metadata = mock.Mock(return_value=None)
-            mock_context._rpc_event.call_details.method = 'hello'
+
+            mock_context._rpc_event.call_details.host = MOCK_HOST
+            mock_context._rpc_event.call_details.method = MOCK_METHOD
```
Comment (lzchen) [inline contrib/opencensus-ext-grpc/tests/test_server_interceptor.py:138]: Update tests to include the additional fields that were added?

File: contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py (hunk 3 above)
Comment (fanshaohua-fan) [reply contrib/opencensus-ext-grpc/opencensus/ext/grpc/server_interceptor.py:161]: The HTTP verbs for gRPC is restricted to POST as described in [gRPC specification](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#L25).

https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#L25

The `grpc_method` is more like the name of the Function.

Here is comparison of gRPC  with HTTP API:
https://docs.microsoft.com/en-us/aspnet/core/grpc/comparison?view=aspnetcore-3.1

File: contrib/opencensus-ext-grpc/tests/test_server_interceptor.py (hunk 5 above)
Comment (fanshaohua-fan) [reply contrib/opencensus-ext-grpc/tests/test_server_interceptor.py:138]: It's already included in the latest test_server_interceptor file.

image

**Comment by liyakun** [comment]
Hi @lzchen, may I ask why this PR has not been merged? Thanks

**Pull request merged**